src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Ricci.cpp

```diff
@@ -0,0 +1,167 @@
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = 0; j < VolumeDim; ++j) {
+        spatial_phi_ijK.get(i, j, k) =
+            0.5 * inverse_spatial_metric.get(k, 0) * phi.get(i, 1 + j, 1);
+        for (size_t l = 1; l < VolumeDim; ++l) {
+          spatial_phi_ijK.get(i, j, k) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(i, 1 + j, 1 + l);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& spatial_phi_trace_second_third_indices =
+      get<::Tags::Tempi<0, VolumeDim, Frame, DataType>>(local_buffer);
+  auto& spatial_phi_trace_first_second_indices =
+      get<::Tags::Tempi<1, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    spatial_phi_trace_second_third_indices.get(i) =
+        spatial_phi_ijK.get(i, 0, 0);
+    spatial_phi_trace_first_second_indices.get(i) =
+        spatial_phi_Ijj.get(0, 0, i);
+    for (size_t j = 1; j < VolumeDim; ++j) {
+      spatial_phi_trace_second_third_indices.get(i) +=
+          spatial_phi_ijK.get(i, j, j);
+      spatial_phi_trace_first_second_indices.get(i) +=
+          spatial_phi_Ijj.get(j, j, i);
+    }
+  }
+
+  auto& spatial_phi_trace_second_third_indices_I =
+      get<::Tags::TempI<0, VolumeDim, Frame, DataType>>(local_buffer);
+  auto& spatial_phi_trace_first_second_indices_I =
+      get<::Tags::TempI<1, VolumeDim, Frame, DataType>>(local_buffer);
+
+  raise_or_lower_index(make_not_null(&spatial_phi_trace_second_third_indices_I),
+                       spatial_phi_trace_second_third_indices,
+                       inverse_spatial_metric);
+  raise_or_lower_index(make_not_null(&spatial_phi_trace_first_second_indices_I),
+                       spatial_phi_trace_first_second_indices,
+                       inverse_spatial_metric);
+
+  for (size_t i = 0; i < VolumeDim; ++i) {
+    for (size_t j = i; j < VolumeDim; ++j) {  // Symmetry
+      ricci->get(i, j) = 0.;
+      for (size_t k = 0; k < VolumeDim; ++k) {
+        ricci->get(i, j) +=
+            0.5 *
+            (phi.get(k, 1 + i, 1 + j) - phi.get(i, 1 + j, 1 + k) -
+             phi.get(j, 1 + i, 1 + k)) *
+            (2.0 * spatial_phi_trace_first_second_indices_I.get(k) -
+             spatial_phi_trace_second_third_indices_I.get(k));
+        for (size_t l = 0; l < VolumeDim; ++l) {
+          ricci->get(i, j) +=
+              0.25 * inverse_spatial_metric.get(k, l) *
+                  (deriv_phi.get(j, l, 1 + k, 1 + i) +
+                   deriv_phi.get(i, l, 1 + k, 1 + j) -
+                   deriv_phi.get(j, i, 1 + k, 1 + l) -
+                   deriv_phi.get(i, j, 1 + k, 1 + l) +
+                   deriv_phi.get(k, i, 1 + l, 1 + j) +
+                   deriv_phi.get(k, j, 1 + l, 1 + i) -
+                   2.0 * deriv_phi.get(l, k, 1 + i, 1 + j)) +
+              spatial_phi_ijK.get(i, k, l) * spatial_phi_ijK.get(j, l, k) +
```

Comment:
the final three terms appear to be different from the documentation (the factor of 0.25 doesn't seem to be applied to them, where there is a 1/4 in the documentation).
It appears that the python implementation shares the discrepancy.

Reply:
thats because the mixed-indexed versions of `spatial_phi` are precomputed with a factor of 1/2. I've updated the help text to clarify this, see the last line.